processing/src/main/java/org/apache/druid/query/metadata/metadata/ColumnAnalysis.java

```diff
@@ -136,7 +147,19 @@ public ColumnAnalysis fold(ColumnAnalysis rhs)
     }
 
     if (!type.equals(rhs.getType())) {
-      return ColumnAnalysis.error("cannot_merge_diff_types");
+      return ColumnAnalysis.error(
+          StringUtils.format("cannot_merge_diff_types: [%s] and [%s]", type, rhs.getType())
+      );
+    }
+
+    if (!typeSignature.equals(rhs.getTypeSignature())) {
```

Comment:
This'll NPE if `typeSignature` is null — i.e. if it came from a server that is on an older version. I think this is OK, since people are supposed to update Brokers after segment-serving-servers, but I'd still feel better if this code was resilient to null `typeSignature`.

Which reminds me: I think we'll also need a trivial change in the cache key to make sure SegmentAnalysis objects from shared caches are not re-used.